whd/dispatcher.py

```diff
@@ -213,6 +219,130 @@ def _set_pr_labels(self, pr_event, resources):
+        '''
+        check all PR resource versions for which no build exists. If a resource version has
+        no build, pin this resource version, trigger the corresponding job and unpin the resource
+        '''
+        for job, resource, resource_version in self._resource_versions_and_jobs(
+            concourse_api, matching_resources, pr_event
+        ):
+            logger.info(
+                f"Pin resource version {resource_version.version()} "
+                f"and trigger build for job {job.name}"
+            )
+            self._trigger_build_with_pinned_resource_version(
+                job=job,
+                resource_name=resource.name,
+                resource_version_id=resource_version.id()
+            )
+
+    def _resource_version_ids_not_built(self, resource_versions, builds):
+        '''
+        yield resource versions which have not been built by checking if the resource version
+        reference is found in an existing build
+        '''
+        for resource_version in resource_versions:
+            resource_version_in_build = False
+            for build in reversed(builds[-30:]):
+                build_plan = build.plan()
+                if build_plan.contains_version_ref(resource_version.version()['ref']):
+                    resource_version_in_build = True
+                    break
+            if not resource_version_in_build:
+                yield resource_version
+
+    def _trigger_build_with_pinned_resource_version(self, job, resource_name, resource_version_id):
+        '''
+        pin the passed resource version, trigger the corresponding job and unpin resourcce
+        version again
+        '''
+        def _job_running(
+            pipeline_name, job_name, build_number, concourse_api, retries=10, sleep_sec=3
+        ):
+            time.sleep(sleep_sec)
+            retries -= 1
+            if retries < 0:
+                logger.warning(
+                    f"Job {job_name} with build number {build_number} cannot be started"
+                )
+                return False
+
+            build = concourse_api.job_build(
+                pipeline_name=pipeline_name,
+                job_name=job_name,
+                build_name=build_number,
+            )
+            if build.status() == BuildStatus.RUNNING:
+                return True
+            else:
+                return _job_running(
+                            pipeline_name,
+                            job_name,
+                            build_number,
+                            concourse_api,
+                            retries,
+                            sleep_sec=sleep_sec*1.2
+                        )
+
+        concourse_api = job.concourse_api
+        pipeline_name = job.pipeline.name
+        version_pin_lock = threading.Lock()
+        with version_pin_lock:
+            concourse_api.pin_resource_version(
```

Comment:
as already discussed earlier:

- check whether the version is already pinned
  - if so: is it already pinned to _our_ version
    - if so: no pinning required - just trigger
  - if not: wait for the version to be unpinned again
   - do this using try-sleep-retry.... for some 1..2m maybe
   - however, check whether the version was maybe pinned by someone else than us (a human..)
    - if so, there is no point in waiting - just bail out
    - if so, we should have also written a comment (in particular: add a timestamp)
      - check the timestamp: if it is too old (say 5m+), unpin it (probably this was a leftover / stems from a bug)

Also, as discussed earlier, there must be a transactional semantics for the resource-pinning:

Use (preferrably) a context mgr. Or - if you do not like using context mgrs, for whichever reason - use a try-finally (which is semantically equivalent).

If the resource is not yet pinned (see above for cases where it is):

- pin it
- add a machine-readable note to it, stating our username, maybe, and a comment (e.g. auto-pinning for PR xyz, ..) and a timestamp ("now")
- try to run the job(s) - there could be more than one
- do this with a timeout
- if either all jobs were successfully triggered, or the timeout expires:
- unpin the version again